### PR TITLE
Add merge sort & quick sort

### PR DESCRIPTION
**Pull request by andreaguarino**

### CODE REVIEW

File: src/sorting/quick_sort.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        if i == j {
+            return j;
+        }
+        list.swap(i, j);
+    }
+}
+
+pub fn _quick_selection<T: Ord + Clone>(list: &mut [T], k: usize) -> &T {
+    if list.len() == 1 {
+        return &list[0];
```
Comment (antonioaversa) [inline src/sorting/quick_sort.rs]: If `list` has length 1 and `k > 0`, shouldn't `_quick_selection` return something like a null value?

Any singleton list containing the value `v` only has a 0-rank item (namely `v`) and no item of any other rank.

File: src/sorting/mod.rs [hunk 2]
```diff
@@ -98,6 +166,10 @@ mod tests {
         let mut singleton_array: Vec<i32> = vec![42];
         sorting_function(&mut singleton_array);
         assert_eq!(singleton_array, vec![42]);
+
+        let mut longer_list = vec![4, 7, 11, 0, 3, 5, 42, 12];
```
Comment (antonioaversa) [inline src/sorting/mod.rs:217]: Can we also have a test with duplicated items?

And possibly also one with all items equal to each other (e.g. `[0, 0, 0, 0]`).

The first would test partition correctness in presence of duplicates, the second would test edge cases in pointers crossing in Hoare partitioning scheme.

**Review by antonioaversa — CHANGES_REQUESTED** [review]
Behavior of quick selection with `k > 0` seems incorrect to me.

The rest looks good.
